### ISSUE TITLE
Backend login redirects to front end login

### ISSUE THREAD

**Issue by Hambern** [rainlab/user-plugin#258]
When I try to login into the backend while not being logged in i the front end. I, after having entered the backend login information, gets redirected to the front end login form. Anyone else experienced this?

**Expected behavior**

If I'm not logged in via the Rainlab.User plugin and tries to login to the backend I expect to get redirected to the dashboard upon login.

**Actual behavior**

I get redirected to /favicon.ico instead. But if I first login into the frontend, this all works.

**Reproduce steps**

Install Rainlab.User, do not log in, visit the backend login, login.

**Comment by LukeTowers** [comment]
I could have sworn somebody reported this exact issue before and then closed it. Let me check.

**Comment by LukeTowers** [comment]
Ah, I see: https://github.com/octobercms/october/issues/2754 it was you but on the main repository.

**Comment by Hambern** [comment]
Exactly it was me :) But on the wrong place ;) I guess it fits better here. I mainly wonder if it's my fault or a general bug

**Comment by LukeTowers** [comment]
I'm not sure, I don't have the time to try to replicate it myself right now.

**Comment by webmaxx** [comment]
It's true.

**Comment by LukeTowers** [comment]
@webmaxx so you can replicate this issue?

**Comment by webmaxx** [comment]
Yes.

### My layouts:

**main.htm**
>>description = "Шаблон главной страницы"
>>
>>[session]
>>security = "user"
>>redirect = "auth/login"
>==
>>html code

**auth.htm**
>>description = "Шаблон страницы для авторизации и т.д."
>>
>>[session]
>>security = "guest"
>>redirect = "index"
>==
>>html code

### My pages:

**index.htm**
>>title = "Главная страница"
>>url = "/"
>>layout = "main"
>>is_hidden = 0
>==
>>html code

**auth/login.htm**
>>title = "Войти"
>>url = "/auth/login"
>>layout = "auth"
>>description = "Страница авторизации"
>>is_hidden = 0
>>
>>[account formLogin]
>>paramCode = "code"
>==
>>{% component 'formLogin' %}

### Steps:

1. Go to "http://site.com". Result - redirect to "http://site.com/auth/login"
2. Go to "http://site.com/backend". Result - redirect to "http://site.com/backend/backend/auth/signin"
3. Submit auth form. Result - redirect to "http://site.com/favicon.ico" + redirect to "http://site.com/auth/login"
4. Go to "http://site.com/backend". Result - profit. I'm authorized

**Comment by Hambern** [comment]
Exactly like that 👍

**Comment by afixia** [comment]
I'm having the same issue. I just made a plugin component that extends the **RainLab\User\Components\Account** component.

Example of using onCustomSignin()

```html
// login.htm
[customaccount]
redirect = "account"
==
<form data-request="onCustomSignin" >
...

```

```php
// component: 'customaccount'
use RainLab\User\Components\Account;
class CustomAccount extends Account
{
  public function onCustomSignin()
  {
    try {
      parent::onSignin();

      $redirectUrl = null;

      // user defined component property
      if ($redirect_value = $this->property('redirect')) {
        $redirectUrl = $this->pageUrl($redirect_value);
      }

      // give preference to request input, otherwise use user defined component property
      $redirectUrl = input('redirect', $redirectUrl);

      // prefer not to use Redirect::intended if we have a user defined url
      if ($redirectUrl) {
        return \Redirect::to($redirectUrl);
      }

      // current method as fallback
      return \Redirect::intended($redirectUrl);

    } catch (\Exception $ex) {
      if (\Request::ajax()) throw $ex;
      else \Flash::error($ex->getMessage());
    }
  }
}
```

**Comment by marthib** [comment]
#274

**Comment by neilcarpenter1** [comment]
This happens the other way around to.

If I log into the backend and then log out.

And then login through the frontend... I am logged in on the front end, but am redirected to the backend login form.

It seems to be line 160 in the Account.php component, using...
`return Redirect::intended($redirectUrl);`
If this is replaced with..
`return Redirect::to($redirectUrl);`

Then you are redirected to your specified redirect url. I'm not sure if there's any specific reason why intended() method is used instead of to();

Is the only current way around this to overwrite on the onSignin method of the component?

**Comment by LukeTowers** [comment]
We can merge that change if necessary, but the reason that `intended()` is used is because it is supposed to redirect the user to the page they were trying to access before they got booted to the login page. This is mostly used by the Session component. If we could figure out why the `intended()` method is picking up the wrong URL to send people to, that would be the ideal solution.

**Comment by neilcarpenter1** [comment]
In my case at least, it will only redirect to the property set in your component if there isn't a `url.intended` session variable set, otherwise it will do that. Which I'm guessing is getting set at some point during the logging out process on the backend.

So to circumvent this, I've overwritten the `onSignin()` method in the code section on my page and cleared the `url.intended` session variable before actually signing in.
`
function onSignin()
{
    try {
    	\Session::forget('url.intended');
        return $this->account->onSignin();
    }
    catch (Exception $ex) {
        Log::error($ex);
    }
}
`

**Comment by LukeTowers** [comment]
To fix this and a couple of other related reported issues, I think it would be best if we could solve the root cause of the `url.intended` session variable being set on backend logout, or perhaps looked at using a different session variable for `url.intended` between frontend sessions and backend sessions.

**Comment by LukeTowers** [comment]
Closing as it has been over a month since any activity on this occurred.

**Comment by mjauvin** [comment]
@LukeTowers I'd like to find a clean solution for this one. The only function that actually SETS this session key (url.intended) is in vendor/laravel/framework/src/Illuminate/Routing/Redirector.php:

```
    public function guest($path, $status = 302, $headers = [], $secure = null)
    {
        $this->session->put('url.intended', $this->generator->full());

        return $this->to($path, $status, $headers, $secure);
    }
```

 but I cannot find where this gets called in the backend authentication code when loging out.

Any pointer?

**Comment by LukeTowers** [comment]
@mjauvin perhaps you can throw some debugging statements in there to find out when it gets called?

**Comment by mjauvin** [comment]
In the mean time, here is an alternative patch that can be added to a Plugin to fix the issue:

```
    public function boot()
    {
        \Event::listen('rainlab.user.beforeAuthenticate', function($account, $credentials) {
            \Session::forget('url.intended');
        });
    }
```

Note: this does not fix the problem in the other direction (Logout of a front-end session then login to backend... backend redirected to front-end login form)

**Comment by mjauvin** [comment]
After tinkering in the code, there is no way to do this cleanly but to create a seperate session variable for the front-end (as @LukeTowers suggested)... we'll need to bypass the framework Redirector class (vendor/laravel/framework/src/Illuminate/Routing/Redirector.php) in order to do this, in particular the "guest()" and "intended()" methods since this is where the session variable url.intended gets set and retreieved, respectively.

**Comment by LukeTowers** [comment]
@mjauvin the code that should be preventing this issue from occurring was implemented 4 years ago, however the issue seems to be caused by the fact the url.intended session variable is not properly being removed by Redirect::intended. Are you able to look into that instead?

**Comment by mjauvin** [comment]
If you mean this, it is not working:

```
--- Redirector.php.dist 2018-03-30 19:56:40.079992990 -0400
+++ Redirector.php      2018-03-30 19:59:28.106992970 -0400
@@ -99,6 +99,7 @@
     public function intended($default = '/', $status = 302, $headers = [], $secure = null)
     {
         $path = $this->session->pull('url.intended', $default);
+        $this->session->forget('url.intended');
 
         return $this->to($path, $status, $headers, $secure);
     }
```

**Comment by mjauvin** [comment]
I do have a working patch to rainlab.user that fixes the problem with a seperate session variable, though.

**Comment by LukeTowers** [comment]
@mjauvin what session driver are you using?

**Comment by mjauvin** [comment]
@LukeTowers session driver = file

**Comment by LukeTowers** [comment]
I really would prefer fixing the original issue (`url.intended` not being removed properly from the session) rather than add a workaround here, but if we can't find out why that's happening then I'll consider the proposed workaround.

**Comment by mjauvin** [comment]
Ok, let's summarize what's really happening

- There are two seperate "environments": the frontend and the backend.
- Both environments share a common session, in which the "url.intended" variable can be defined.
- When requesting a page on the frontend or backend which can only be accessed while being logged in, if the user is NOT currently logged in, October will set the "url.intended" session variable to the requested URL and redirect the user to the respective Login form (backend or frontend).
- Once the user is being logged in succesfully, the "url.intended" session variable is retrieved from the session using the "pull" Session Store method, the session variable is forgotten and the user is redirected to the URL retrieved from this session variable (this works fine).

Problems happen when the url.intended session variable gets set within an environment, THEN, before the user is logged into this environment, a login request is made within the OTHER environment which triggers the url.intended variable to get used to redirect the user after successful login... 

Let's give concrete examples:

Scenario #1
- User opens a browser window to http://example.com/backend/cms
- This URL is not accessible without being properly logged in, so the url.intended session variable gets set to "http://example.com/backend/cms" and user gets redirected to "http://example.com/backend/backend/auth/signin"
- User opens another browser window to "http://example.com/login"
- User enters proper login information and gets redirected to the URL currently defined into the url.intended session variable ("http://example.com/backend/cms") <--- PROBLEM!

Scenario #2
- User is currently logged into the backend
- User opens another browser window to "http://example.com/private" (not accessible unless logged into the frontend)
- url.intended is set to "http://example.com/private" and user gets redirected to "http://example.com/login".
- User logs out of the backend session, gets redirected to "http://example.com/backend/backend" (ref. modules/backend/controllers/Auth.php::signout) but that URL redirects to the backend login form and sets the url.intended to  "http://example.com/backend/backend".
- If User goes back to previous browser window and logs into the frontend, he gets redirected to the current url.intended (http://example.com/backend/backend) which is not accessible without being logged into the backend, thus, user gets redirected to the backend login form. <--- PROBLEM!

So the only way to avoid those problems is to use a different session variable for the FRONTEND, as implemented by PR #324 

Thanks.

**Comment by LukeTowers** [comment]
Fixed by @mjauvin in https://github.com/rainlab/user-plugin/commit/2ef30299e22810da42b0cae824a5a63474c39b99, thanks for your effort on finding a fix @mjauvin!
